### ISSUE TITLE
https://eugeneyan.com/writing/recommender-systems-baseline-pytorch/

### ISSUE THREAD

**Issue by eugeneyan** [eugeneyan/eugeneyan-comments#8]
Migrated from json into utteranc.es

**Comment by ityutin** [comment]
Hi, Eugene! Please help me out here: I understood the nature of "Continuous labels" but I don't understand the objective you use for training and how we get from "Continuous labels" to ROC AUC calculations.

**Comment by eugeneyan** [comment]
Hey Ilya, 

TL;DR: While we use continuous labels for training, the output layer is a sigmoid that return scores from 0 - 1.

--- 

_Why use continuous labels?_ Because they're able to better distinguish the strength of product-pair relationships. Here's a breakdown on binary vs. continuous labels. 

**Binary labels**: A pair of products have a label of 1 if they have any, or multiple, relationships (e.g., `also viewed`, `also bought`, `bought together`); zero otherwise. 

This means that a pair of products which only have the `also viewed` will have the same label as a pair of products which have all three relationships (i.e., `also viewed`, `also bought`, `bought together`). While this approach can tell us if a pair of products have, or do not have, a relationship, it doesn't distinguish the strength of the relationship. 

**Continuous labels**: To distinguish between the strengths of relationships, we use the following labels (instead of 1 or 0) - `bought together` = 1.2, `also bought` = 1.0, `also viewed` = 0.5.

So, binary labels have values of 0 or 1.0, continuous labels have values of 0, 0.5, 1.0, 1,2. 

---

_Regardless of the type of label, we get predictions from a simple sigmoid_, and predictions will range from 0 - 1. (See code for matrix factorization on [binary labels](https://github.com/eugeneyan/recsys-nlp-graph/blob/master/src/ml/mf.py#L33) and [continuous labels](https://github.com/eugeneyan/recsys-nlp-graph/blob/master/src/ml/mf_continuous.py#L40)). Nonetheless, we use binary cross entropy loss for binary labels, and mean square error loss for continuous labels. 

Given that the final layer for both binary and continuous labels is a sigmoid, we can use ROC AUC for both.

**Comment by ityutin** [comment]
Wow, such a comprehensive answer. ) 
First off, I'm very sorry that I overlooked the link to the repo 🤦 
All is clear now. I got confused because, for me, regression and classification were like "two different worlds"  due to projects I previously worked on. But I can see it makes sense here.

P.S. Probably it would be a good idea to incorporate your excellent comment into the article to help someone like me in the future

**Comment by eugeneyan** [comment]
Thanks for the idea Ilya! Have added it into the article, right after the precision recall curves for Matrix Factorization (continuous labels).

**Comment by vibhas-singh** [comment]
Hi Eugene,
Great article - really insightful.

Can you throw some more light on what exactly is `cliff of death` and how the second model with bias is more production friendly even when it has lesser precision and recall?

**Comment by dvquy13** [comment]
Hi Eugene, thanks for sharing the project, I have learned a lot!

I have two questions:

1. IIUC you measure AUC-ROC for continuous label by getting the sigmoid predictions and then compare them against the binary versions of the val dataset right? So even though in training dataset our labels are continous, in val dataset we use binary labels so that we can measure AUC-ROC.

2. How does this approach fit in the Retrieval-Ranking phases of a RecSys? Is this a model in the retrieval phase? If yes then at inference step do we just retrieve candidate recommendations based on similarity-lookup on top of the learned embeddings?
